_posts/2018-05-18-contributing-to-installer-puppet-modules.md

```diff
@@ -0,0 +1,326 @@
+---
+layout: post
+title: 'Contributing to the Installer Puppet Modules'
+date: 2018-05-21 12:00:00
+author: Christopher Duryee
+tags:
+- foreman
+---
+---
+layout: post
+title: 'Contributing to the Installer Puppet Modules'
+date: 2018-05-21 12:00:00
+author: Christopher Duryee
+tags:
+- foreman
+---
+
+# How do I get started with contributing to the installer?
+
+The Foreman installer can be intimidating at first. It manages lots of files, and has many, many, many options. Even more options are exposed via `custom-hiera.yaml`. If you want to make a change that's not covered by either an installer option or `custom-hiera.yaml`, you may feel like you are out of luck. However, this is a great oppurtunity to make an upstream contribution!
+
+This blog post shows the steps on how to identify a change you want to make in the installer and create an upstream pull request in the correct installer module.
+
+## Initial setup
+
+*The workflow below describes how I would proceed with an installer change. However, everyone is different and there are multiple ways to go about doing the same thing.*
+
+If you found an issue to fix, it's likely on your production Foreman installation. This is not a suitable place to edit installer code! Do not do this. Instead, use [forklift](https://github.com/theforeman/forklift). Forklift uses Vagrant and Ansible to create development VMs on a hypervisor. I have only used Forklift with Fedora but I've heard it works on Debian and Ubuntu.  You'll need to find a place to run Forklift before continuing.
```

Comment:
Forklift is pretty intensive as it can only create Katello dev setups - maybe we should mention that you can just installer the foreman-installer on a fresh CentOS/Debian host and hack on the modules there, if you need a vanilla install? (side note, we really need to get forklift working for vanilla setups :P)

Reply:
It's true there is no git-based checkout version of a pure Foreman box. But for installer change testing, we tend to use nightly installs and not git-based checkouts. With that in mind, there are lots of options for vanilla Foreman:

```
centos7-foreman               not created (libvirt)
centos7-foreman-1.15          not created (libvirt)
centos7-foreman-1.16          not created (libvirt)
centos7-foreman-1.17          not created (libvirt)
centos7-foreman-bats-ci       not created (libvirt)
centos7-foreman-nightly       not created (libvirt)
centos7-foreman-proxy-3.4     not created (libvirt)
centos7-foreman-proxy-3.5     not created (libvirt)
centos7-foreman-proxy-3.6     not created (libvirt)
centos7-foreman-proxy-nightly not created (libvirt)
debian8-foreman-1.15          not created (libvirt)
debian8-foreman-1.16          not created (libvirt)
debian8-foreman-nightly       not created (libvirt)
debian9-foreman-1.16          not created (libvirt)
debian9-foreman-1.17          not created (libvirt)
debian9-foreman-nightly       not created (libvirt)
ubuntu1604-foreman-1.15       not created (libvirt)
ubuntu1604-foreman-1.16       not created (libvirt)
ubuntu1604-foreman-1.17       not created (libvirt)
ubuntu1604-foreman-nightly    not created (libvirt)
```